ihaskell-display/ihaskell-widgets/src/IHaskell/Display/Widgets/Button.hs

```diff
@@ -0,0 +1,210 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module IHaskell.Display.Widgets.Button (
+    -- * The Button Widget
+    Button,
+    -- * Predefined button styles
+    ButtonStyle(..),
+    -- * Create a new button
+    mkButton,
+    -- * Set button properties
+    setButtonStyle,
+    setButtonLabel,
+    setButtonTooltip,
+    disableButton,
+    enableButton,
+    toggleButtonStatus,
+    -- * Get button properties
+    getButtonStyle,
+    getButtonLabel,
+    getButtonTooltip,
+    isDisabled,
+    -- * Click handlers
+    setClickHandler,
+    getClickHandler,
+    triggerClick,
+    ) where
+
+-- To keep `cabal repl` happy when running from the ihaskell repo
+import           Prelude
+
+import           Control.Monad (when)
+import           Data.Aeson (ToJSON, Value(..), object, toJSON, (.=))
+import           Data.Aeson.Types (Pair)
+import           Data.HashMap.Strict as Map
+import           Data.IORef
+import           Data.Text (Text)
+import qualified Data.Text as T
+import           System.IO.Unsafe (unsafePerformIO)
+
+import           IHaskell.Display
+import           IHaskell.Eval.Widgets
+import qualified IHaskell.IPython.Message.UUID as U
+import           IHaskell.Types (WidgetMethod(..))
+
+-- | ADT for a button
+data Button =
+       Button
+         { uuid :: U.UUID
+         , description :: IORef Text
+         , tooltip :: IORef Text
+         , disabled :: IORef Bool
+         , buttonStyle :: IORef ButtonStyle
+         , clickHandler :: IORef (Button -> IO ())
+         }
+
+-- | Pre-defined button-styles
+data ButtonStyle = Primary
+                 | Success
+                 | Info
+                 | Warning
+                 | Danger
+                 | None
+  deriving (Eq, Show)
+
+-- | Create a new button
+mkButton :: IO Button
+mkButton = do
+  -- Default properties, with a random uuid
+  uuid <- U.random
+  sender <- newIORef Nothing
+  desc <- newIORef "label"        -- Non-empty to get a display
+  ttip <- newIORef ""
+  dis <- newIORef False
+  sty <- newIORef None
+  fun <- newIORef (\_ -> return ())
+
+  let b = Button uuid desc ttip dis sty fun
+
+  -- Open a comm for this widget, and store it in the kernel state
+  widgetSendOpen b (toJSON ButtonInitData) (toJSON b)
+
+  -- Return the button widget
+  return b
+
+update :: Button -> [Pair] -> IO ()
```

Comment:
If `Pair` is from `aeson`, perhaps you could use `FromJSON` directly in these rather than forcing all their users to call `toJSON`?